patroni/validator.py

```diff
@@ -354,6 +354,33 @@ def __init__(self, schema: Dict[str, Any]) -> None:
         self._schema = schema
 
 
+class CaseOptions(object):
+    """Map how a :class:`dict` of optional configurations should be validated."""
+
+    def __init__(self, schema: Dict['Optional', Any]) -> None:
+        """Create a :class:`CaseOptions` object.
+
+        :param schema: the optional arguments which may appear in the configuration.
+                       Each option may or may not be provided, and will work with :class:`OneOf`
+                       or similar validations.
+
+        :Example:
+
+            >>> s = Schema({OneOf("nofailover", "failover_priority"): CaseOptions({
+            ...   Optional("nofailover"): bool,
+            ...   Optional("failover_priority"): IntValidator(min=0, raise_assert=True),
+            ... })})
+            ...
+            >>> s({'nofailover': True})
+            []
+            >>> s({'nofailover': True, 'failover_priority': 0})
+            ["  Multiple of ('nofailover', 'failover_priority') provided"]
+
+        That will make sure that at most one of ``nofailover`` or ``failover_priority`` is provided.
+        """
+        self._schema = {k.name: v for k, v in schema.items()}
+
+
```

Comment:
Do we really need this new class?

I do see value in `OneOf`, but I think we could just the values like this -- using `Case` instead of `CaseOptions` + `Optional`:

```python
        OneOf("nofailover", "failover_priority"): Case({
            "nofailover": bool,
            "failover_priority": IntValidator(min=0, raise_assert=True),
        }),
```

The result would be apparently the same when running `--validate-config`, and the `OneOf` name already gives me the idea that only one of the options should be set.

Just to be clear, something like this is what I propose:

```diff
diff --git a/patroni/validator.py b/patroni/validator.py
index 3d0c1d1..fca2657 100644
--- a/patroni/validator.py
+++ b/patroni/validator.py
@@ -354,33 +354,6 @@ class Case(object):
         self._schema = schema


-class CaseOptions(object):
-    """Map how a :class:`dict` of optional configurations should be validated."""
-
-    def __init__(self, schema: Dict['Optional', Any]) -> None:
-        """Create a :class:`CaseOptions` object.
-
-        :param schema: the optional arguments which may appear in the configuration.
-                       Each option may or may not be provided, and will work with :class:`OneOf`
-                       or similar validations.
-
-        :Example:
-
-            >>> s = Schema({OneOf("nofailover", "failover_priority"): CaseOptions({
-            ...   Optional("nofailover"): bool,
-            ...   Optional("failover_priority"): IntValidator(min=0, raise_assert=True),
-            ... })})
-            ...
-            >>> s({'nofailover': True})
-            []
-            >>> s({'nofailover': True, 'failover_priority': 0})
-            ["  Multiple of ('nofailover', 'failover_priority') provided"]
-
-        That will make sure that at most one of ``nofailover`` or ``failover_priority`` is provided.
-        """
-        self._schema = {k.name: v for k, v in schema.items()}
-
-
 class Or(object):
     """Represent the list of options that are available.

@@ -425,15 +398,14 @@ class Optional(object):


 class OneOf(object):
-    """Mark that at most one option from a `CaseOptions` can be suplied
+    """Mark that at most one option from a `Case` can be suplied

     Represents a list of possible configuration options in a given scope, where at most one can actually
     be provided.

     .. note::

-        It should be used together with a :class:`CaseOptions` object, where every field in the :class:`CaseOptions`
-        is something that is :class:`Optional`
+        It should be used together with a :class:`Case` object.
     """

     def __init__(self, *args: str) -> None:
@@ -443,9 +415,9 @@ class OneOf(object):

         :Example:

-            OneOf("nofailover", "failover_priority"): CaseOptions({
-                Optional("nofailover"): bool,
-                Optional("failover_priority"): IntValidator(min=0, raise_assert=True),
+            OneOf("nofailover", "failover_priority"): Case({
+                "nofailover": bool,
+                "failover_priority": IntValidator(min=0, raise_assert=True),
             })

         The :class`OneOf` object is used to define that at most one of ``nofailover`` and
@@ -732,7 +704,7 @@ class Schema(object):
                         self.data[d] = key.default
                     validator = self.validator[key]
                     if isinstance(key, Or) and isinstance(self.validator[key], Case) \
-                            or isinstance(key, OneOf) and isinstance(self.validator[key], CaseOptions):
+                            or isinstance(key, OneOf) and isinstance(self.validator[key], Case):
                         validator = self.validator[key]._schema[d]
                     # In this loop we may be calling a new `Schema` either over an intermediate node in the tree, or
                     # over a leaf node. In the latter case the recursive calls in the given path will finish.
@@ -1115,9 +1087,9 @@ schema = Schema({
         Optional("safety_margin"): int
     },
     Optional("tags"): {
-        OneOf("nofailover", "failover_priority"): CaseOptions({
-            Optional("nofailover"): bool,
-            Optional("failover_priority"): IntValidator(min=0, raise_assert=True),
+        OneOf("nofailover", "failover_priority"): Case({
+            "nofailover": bool,
+            "failover_priority": IntValidator(min=0, raise_assert=True),
         }),
         Optional("clonefrom"): bool,
         Optional("noloadbalance"): bool,
```

Reply:
I like this better! Updated.

Reply:
Cool!